### PR TITLE
Android twilio v3

### PR DESCRIPTION
**Pull request by fabriziomoscon**
This PR is a stepping stone to upgrade Android to Twilio Voice SDK v5.
On top of the current master (59304db) changes include the following PRs:
- #108 (by @fabriziomoscon)
- #130 (by @Pagebakers) 

+ additional changes needed to complete implementation of Twilio Voice SDK v3.3.0
- https://github.com/twilio/voice-quickstart-android/pull/270/files#diff-bc72a3560f9dc9b186a86db4602f5d1d
- https://github.com/twilio/voice-quickstart-android/blob/master/Docs/new-features-3.0.md

experiment for Android Twilio Voice SDK v3.3.0
- add hold() call
- add ringing state
- remove state for callInvite
- handle internal CANCEL_CALL notification
- add ConnectOptions
- add AcceptOptions
- integrate the removal of callInvite of state PENDING

TODO Android
- [x] implement JS method to send ConnectOptions
- [x] test Android notification for incoming calls: app running in foreground
- [x] test Android notification for incoming calls: app running in background
- [x] test Android notification for incoming calls: app not running
- [x] test Android outgoing calls: app running in foreground
- [x] test on JS app
- [x] add additional feature into README
- [x] consider https://github.com/hoxfon/react-native-twilio-programmable-voice/pull/142
- [x] implement Android v4
- [x] fix: https://github.com/hoxfon/react-native-twilio-programmable-voice/pull/144#issuecomment-595260805
- [ ] implement Android v5
- [ ] fix https://github.com/hoxfon/react-native-twilio-programmable-voice/pull/144#issuecomment-596589874

TODO iOS
Consult this: https://www.twilio.com/docs/voice/voip-sdk/ios/changelog#300
- [x] https://github.com/twilio/voice-quickstart-ios/blob/master/Docs/migration-guide-2.x-3.x.md
- [x] https://github.com/twilio/voice-quickstart-ios/blob/master/Docs/migration-guide-3.x-4.x.md
- [x] include #138
- [x] test iOS notification for incoming calls: app running in foreground
- [x] test iOS notification for incoming calls: app running in background
- [x] test iOS notification for incoming calls: app not running
- [x] test iOS outgoing calls: app running in foreground
- [ ] test on JS app
- [x] update CHANGELOG with iOS updates
- [ ] reproduce on iOS 13: https://github.com/hoxfon/react-native-twilio-programmable-voice/pull/144#issuecomment-595278001
- [ ] format all the code as a separate commit according to latest XCode style

### CODE REVIEW

File: android/src/main/java/com/hoxfon/react/RNTwilioVoice/TwilioVoiceModule.java [hunk 1]
```diff
@@ -473,11 +515,14 @@ private void handleIncomingCallIntent(Intent intent) {
...
+                Log.d(TAG, "VoiceBroadcastReceiver.onReceive "+action+". Intent "+ intent.getExtras());
+            }
             if (action.equals(ACTION_INCOMING_CALL)) {
-                if (BuildConfig.DEBUG) {
-                    Log.d(TAG, "VoiceBroadcastReceiver.onReceive ACTION_INCOMING_CALL. Intent "+ intent.getExtras());
-                }
                 handleIncomingCallIntent(intent);
+            } else if (action.equals(ACTION_CANCEL_CALL_INVITE)) {
+                // TODO check if implementation need something else
+                activeCallInvite = null;
```
Comment (fabriziomoscon) [inline android/src/main/java/com/hoxfon/react/RNTwilioVoice/TwilioVoiceModule.java]: consider using this `clearIncomingNotification(activeCallInvite);`

**Comment by hanipcode** [comment]
Hi, thank you so much for maintaining this library. I really need this update as I am currently working on an App that depends on this library.
I don't have any native knowledge but I can help with testing / bug reproducing. thanks!

**Comment by fabriziomoscon** [comment]
Hi, you can test the Android part by adding this branch:
`yarn add https://github.com/hoxfon/react-native-twilio-programmable-voice#android-twilio-v3`
I am still working on the iOS part

**Comment by fabriziomoscon** [comment]
Hi guys, I pushed a version to review in this branch which includes Android SDK v4 and iOS SKD v5.
It would greatly appreciated if you could test it: @Pagebakers @gitstud @mysport12 @lonnylot @aniravi24 @jdegger

**Comment by fabriziomoscon** [comment]
On iOS the assertion inside the method `provider:performAnswerCallAction` if failing when accepting a call:
```bash
2020-02-16 19:15:29.539111+0000 Hoxfon[532:44731] *** Assertion failure in -[RNTwilioVoice provider:performAnswerCallAction:], /Users/fab/projects/hoxfon/client/node_modules/react-native-twilio-programmable-voice/ios/RNTwilioVoice/RNTwilioVoice.m:625
2020-02-16 19:15:29.543008+0000 Hoxfon[532:44731] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'We only support one Invite at a time.'
```

**Comment by jdegger** [comment]
Hi @fabriziomoscon, unfortunately we do not have any resources available to test this this week. If you do not have the desired test results in somewhere next week we can look to allocate some time!

**Comment by Pagebakers** [comment]
Getting Failed to resolve: support-v4 in react-native-twilio-programmable-voice/android/build.gradle

**Comment by Pagebakers** [comment]
I managed to upgrade the original project after all. First test calls are working. Working on rewriting a lot of code to new apis now and then will test it on a real device.

**Comment by Pagebakers** [comment]
Android seems to be working fine! Next challenge is to get IOS running.

**Comment by fabriziomoscon** [comment]
@Pagebakers how did you resolve `Failed to resolve: support-v4 in react-native-twilio-programmable-voice/android/build.gradle`? Is there anything there that we can learn or perhaps add a note to the migration in the README?

**Comment by Pagebakers** [comment]
I didn't fully move my project to androidx, which caused these issues.

**Comment by jdegger** [comment]
@Pagebakers using https://github.com/mikehardy/jetifier might make it easier, we run this as a npm `postinstall` script

**Comment by Pagebakers** [comment]
Thanks @jdegger Yes jetifier is enabled once you upgrade to 0.60+, this helped to get the older packages back to work. but i had to remove and re-add them one by one to get the app to fully build again.

**Comment by fmonsalvo** [comment]
@fabriziomoscon are you maintaining this repo now? Is there anything I can do to help out?

**Comment by fabriziomoscon** [comment]
Hi @fmonsalvo yes I am maintaining this library. All my efforts are in this PR. I would really appreciate any help especially for the iOS code. Currently I am working to fix this issue: https://github.com/hoxfon/react-native-twilio-programmable-voice/pull/144#issuecomment-586744506

**Comment by fabriziomoscon** [comment]
I pushed a change that removes the above issue by removing the assertion...
this line: 
https://github.com/twilio/voice-quickstart-objc/blob/5.0.0/ObjcVoiceQuickstart/ViewController.m#L534
is removed in the combined repo:
https://github.com/twilio/voice-quickstart-ios/blob/master/ObjcVoiceQuickstart/ViewController.m#L541

**Comment by fabriziomoscon** [comment]
@Pagebakers @gitstud @mysport12 @lonnylot @aniravi24 @jdegger @fmonsalvo
I have updated the iOS part to the latest changes from https://github.com/twilio/voice-quickstart-ios
You can now update your pods to `pod 'TwilioVoice', '~> 5.2.0'`

Let me know if you find some issue while testing.

**Comment by fmonsalvo** [comment]
@fabriziomoscon I'll be testing today both android and iOS, I'll let you know if I find something is not working.

**Comment by fmonsalvo** [comment]
@fabriziomoscon Have you tested the android implementation against android 10? I'm having an issue with incoming calls and the app in the foreground or with the phone locked.

Since Android 10, you cannot start an activity from the background anymore, so the code in `VoiceFirebaseMessagingService.onMessageReceived` when the app is in the background stopped working. I confirmed that sending a notification with full screen intent instead of `context.startActivity(launchIntent);` solves the issue for me.
If you confirm the issue is happening on your side I can add a fix for this situation.

**Comment by fmonsalvo** [comment]
@fabriziomoscon just tested the iOS part and is working fine in most cases, but there's an issue (at least I'm having it) when the app is killed.
When I call a phone with the application killed there's no incoming call action, I don't see anything. Looking at the device console I get this error: `Killing VoIP app <private> because it failed to post an incoming call in time.`
I have been in talks with the team at Twilio trying to solve this issue in my fork of the library, but have not gotten far (https://github.com/twilio/voice-quickstart-ios/issues/308).
Can you confirm this is working on your side?

**Comment by Pagebakers** [comment]
I have some trouble that the connection gets disconnected on Android sometimes, within the first 20 seconds. I haven't been able to figure out the issue. anybody else having similar issues?

**Comment by fmonsalvo** [comment]
@Pagebakers It may be related to this https://github.com/twilio/voice-quickstart-android/issues/178.
It has been fixed in SDK v5.

**Comment by Pagebakers** [comment]
> @Pagebakers It may be related to this [twilio/voice-quickstart-android#178](https://github.com/twilio/voice-quickstart-android/issues/178).
> It has been fixed in SDK v5.

This makes sense, thanks!

We actually ring multiple devices and cancel outstanding devices (js client + mobile sdk) once one device connects.

**Comment by fabriziomoscon** [comment]
@fmonsalvo in regards to this: https://github.com/hoxfon/react-native-twilio-programmable-voice/pull/144#issuecomment-595278001
I have tested on real device iPhone6 iOS 12.4.5 and I can receive calls when the device is locked and the app is killed (is killed the right term? I mean it is not in foreground nor in background).
I collected the logs from XCode device console here: https://gist.github.com/fabriziomoscon/165c16f3a89d9b6b8bccd5d4a5337dec

Note that - my understanding is - you can only receive incoming calls for a build not released by Apple (AppStore and TestFlight) by using Sandbox SDK certificates (Sandbox checkbox ticked on Twilio console). You can find in the log this: `sandbox.push.apple.com`, check if you can see that in the log before the Killing message.

**Comment by fmonsalvo** [comment]
Thanks @fabriziomoscon it looks like it's something on my side of things. I'll keep working on it and I'm glad is not a problem with the library as this will help move this forward.

Can you confirm the behaviour I mentioned on Android 10?

**Comment by Pagebakers** [comment]
Trying IOS now, but i keep getting module TwilioVoice not found. in RNTWilioVoice.m

I've installed 5.2.0 with pods, the framework is included, i can't figure it out. anyone had the same issue?

**Comment by fmonsalvo** [comment]
Is the message `RNTWilioVoice.m` ? Cause that would be a typo since the file name is `RNTwilioVoice.`. I believe xcode and xcodebuild are case sensitive.

**Comment by Pagebakers** [comment]
No sorry for the typo, the error is thrown inside the RNTwilioVoice.m class, the error is "module TwilioVoice not found"

**Comment by enterteg** [comment]
Any plans to migrate to v5 soon? :) Great job BTW!

**Comment by aniravi24** [comment]
We haven't been working on the app that uses this particular library anymore, so we haven't been able to test. Sorry about that! A couple notes based on comments I see here:
1. If iOS is killing your app, it might have to do with the pushkit policy issue described here: https://github.com/twilio/twilio-voice-ios/blob/Releases/iOS-13-Migration-Guide.md#ios-13--xcode-11

2. On Android, as mentioned above, you can't start an activity from the background anymore unless it fits within specific criteria. We used react-native-callkeep (https://github.com/react-native-webrtc/react-native-callkeep) and wrote a small broadcast receiver in our app that would take a broadcast from this library (we patched it to send a broadcast when a call notification came in), and then render the native Android call UI, like CallKit does on iOS. It took more work, but it was worth it.
More generally, the flow was
RN Twilio library receives notification -> sends broadcast to our app -> broadcast receiver moves data to an Android service that is one of the headless services in React Native -> service sends info to JS layer -> RN CallKeep handles the rest once the call information is at the JS layer.
This flow is specifically for when calls happen in the background and the app isn't in view.

**Comment by jdegger** [comment]
@fabriziomoscon I have time to perform some tests right now, can you point me in the direction of what you want me to test specifically?

**Comment by jdegger** [comment]
@fabriziomoscon any updates? We are still willing to dedicate some time.

**Comment by mysport12** [comment]
@fmonsalvo Have you had any luck with the app termination issue on iOS13? I looked at the issue you raised with Twilio and tried implementing some of the suggestions made there to no avail. This is definitely a big issue for us

**Comment by fmonsalvo** [comment]
Hi @mysport12 no luck yet, I reviewed other voip libraries, such as vox-implant's and then I read this comment: https://github.com/react-native-webrtc/react-native-voip-push-notification#important-note.
But I've still to solve the issue. Right now my app is working, but this problem is blocking me from new releases. If you find a solution please drop a line!

**Comment by mysport12** [comment]
@fmonsalvo Ok good to know and I will definitely post back here should I find a solution.

**Comment by josaric** [comment]
iOS CallKit is not showing up for me, so I have custom CallKit created, but I can't accept incoming call on iOS

Please take a look at this link: https://www.twilio.com/docs/voice/client/tutorials/incoming-calls?code-sample=code-accept-an-incoming-twilio-client-call-3&code-language=Objective-C&code-sdk-version=default

I'm using branch #android-twilio-v3

**Comment by jdegger** [comment]
We encounter the following problem on this branche when setting speaker to true

```
Thread 1: EXC_BAD_ACCESS (code=1, address=0x10)
```

Happening here:
```c
#pragma mark - AVAudioSession
- (void)toggleAudioRoute:(BOOL)toSpeaker {
    // The mode set by the Voice SDK is "VoiceChat" so the default audio route is the built-in receiver.
    // Use port override to switch the route.
    self.audioDevice.block =  ^ {
        // We will execute `kDefaultAVAudioSessionConfigurationBlock` first.
        kTVODefaultAVAudioSessionConfigurationBlock();

        // Overwrite the audio route
        AVAudioSession *session = [AVAudioSession sharedInstance];
        NSError *error = nil;
        if (toSpeaker) {
            if (![session overrideOutputAudioPort:AVAudioSessionPortOverrideSpeaker error:&error]) {
                NSLog(@"Unable to reroute audio: %@", [error localizedDescription]);
            }
        } else {
            if (![session overrideOutputAudioPort:AVAudioSessionPortOverrideNone error:&error]) {
                NSLog(@"Unable to reroute audio: %@", [error localizedDescription]);
            }
        }
    };
    self.audioDevice.block();
}
```

On the last line:
```c
 self.audioDevice.block();
```

**Comment by jdegger** [comment]
Our main findings are currently:
- Android is working mostly as expected, we have not tested all use cases yet and we are integrating with CallKeep to overcome the problems encountered on Android 10. We are following @aniravi24's https://github.com/hoxfon/react-native-twilio-programmable-voice/issues/136#issuecomment-617991051 which also works in this branch.
- on iOS we are not seeing good results so far. The crash I reported yesterday is not the only thing causing problems. The call does not seem to connect.

**Comment by jdegger** [comment]
Revoking above iOS crash and findings. It seems to work just fine on iOS so far. Our implementation with CallKeep on iOS was causing trouble.

**Comment by Pagebakers** [comment]
Finally found some time to test again.

Android seems to work fine, except for calls not being closed when calling Twilio.disconnect()

Anybody else experiencing this?

**Comment by Pagebakers** [comment]
> Trying IOS now, but i keep getting module TwilioVoice not found. in RNTWilioVoice.m
> 
> I've installed 5.2.0 with pods, the framework is included, i can't figure it out. anyone had the same issue?

I had to update the Framework Search Paths "$(SRCROOT)/../../../ios/Pods/TwilioVoice" in RNTwilioVoice.xcodeproj to recursive for it to work. The Pod location has changed.

**Comment by jdegger** [comment]
I think we already had it on recursive before hence why we didn't encounter that problem

**Comment by jdegger** [comment]
iOS bug in this branche:
- `device.proximityMonitoringEnabled = NO;` is not called when the call is ended through CallKit, it works fine when you call `disconnect()` yourself from the app.

File: ios/RNTwilioVoice/RNTwilioVoice.m [hunk 2]
```diff
@@ -555,7 +734,6 @@ - (void)performEndCallActionWithUUID:(NSUUID *)uuid {
   if (uuid == nil) {
     return;
   }
-
   UIDevice* device = [UIDevice currentDevice];
   device.proximityMonitoringEnabled = NO;
```
Comment (jdegger) [inline ios/RNTwilioVoice/RNTwilioVoice.m:738]: I had to move this to `didDisconnectWithError` or `callDisconnected` for this to work. I'm not sure this is the correct place for the call.

**Comment by fmonsalvo** [comment]
@jdegger @Pagebakers are you testing with iOS 13 and above?
Have you ran into this issue: https://github.com/twilio/voice-quickstart-ios/issues/308 ?
I will be testing this library today and I'll report my findings.
Thanks.

**Comment by jdegger** [comment]
We have not finished elaborate incoming call tests for iOS. We mostly work with outgoing calls. Incoming calls are on the shortterm roadmap though.

We test on ios 13 and 14

**Comment by calz10** [comment]
Hi Folks!, do we have updates regarding this upgrades? thanks watching here.

**Comment by Pagebakers** [comment]
> 2\. https://github.com/react-native-webrtc/react-native-callkeep

@aniravi24 Do you have example code how you set up this broadcaster in Android? would be great to get callkeep working with this.

**Comment by aniravi24** [comment]
@Pagebakers I believe the comment I made here answers your question, please let me know if it doesn't! https://github.com/hoxfon/react-native-twilio-programmable-voice/issues/136#issuecomment-617991051

**Comment by jdegger** [comment]
@Pagebakers @aniravi24 bit messy but the implementation is also viewable here, this works quite well for Android. We just started looking into implementing iOS again.

https://github.com/hoxfon/react-native-twilio-programmable-voice/compare/android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46

Thanks again to @aniravi24

**Comment by ealexseen** [comment]
Hello everyone, when can I use this update? Thanks you.

**Comment by calz10** [comment]
hi all, do you think when can we merge this PR please? our prod app is relying on this wrapper and i cant find other package. :( thank you for the updates! :) @jdegger

**Comment by jdegger** [comment]
@calz10 I'm not the maintainer of this repository / plugin. That's @fabriziomoscon. So far it's been a bit silent on his side for a while. I have tried reaching out to him to discuss continuing his work on this plugin, so far no luck. 

We privately offered him a bounty. If other companies are also willing to pitch in an support the development of this library we might be able to setup a fund to sponsor this plugin. 

@fabriziomoscon Could you maybe share if you have any time dedicated to this project or would like for others to take over?

**Comment by Pagebakers** [comment]
I've got everything running on Android and IOS in development mode.

But archiving IOS now results in this:
```
Undefined symbols for architecture armv7:
  "_OBJC_CLASS_$_TVOAcceptOptions", referenced from:
      objc-class-ref in RNTwilioVoice.o
  "_OBJC_CLASS_$_TVOConnectOptions", referenced from:
      objc-class-ref in RNTwilioVoice.o
  "_OBJC_CLASS_$_TwilioVoice", referenced from:
      objc-class-ref in RNTwilioVoice.o
  "_OBJC_CLASS_$_TVODefaultAudioDevice", referenced from:
      objc-class-ref in RNTwilioVoice.o
ld: symbol(s) not found for architecture armv7
```

Can't figure out a solution, any ideas?

**Comment by mysport12** [comment]
> I've got everything running on Android and IOS in development mode.
> 
> But archiving IOS now results in this:
> 
> ```
> Undefined symbols for architecture armv7:
>   "_OBJC_CLASS_$_TVOAcceptOptions", referenced from:
>       objc-class-ref in RNTwilioVoice.o
>   "_OBJC_CLASS_$_TVOConnectOptions", referenced from:
>       objc-class-ref in RNTwilioVoice.o
>   "_OBJC_CLASS_$_TwilioVoice", referenced from:
>       objc-class-ref in RNTwilioVoice.o
>   "_OBJC_CLASS_$_TVODefaultAudioDevice", referenced from:
>       objc-class-ref in RNTwilioVoice.o
> ld: symbol(s) not found for architecture armv7
> ```
> 
> Can't figure out a solution, any ideas?

Under Build Setting remove that architecture from VALID_ARCHS, clean, and rebuild.

**Comment by Pagebakers** [comment]
@mysport12 Thanks yeah i found this. But i can find VALID_ARCHS in my build settings :(

**Comment by mysport12** [comment]
@Pagebakers for me in Xcode 12 it was under User Defined Settings. Hope that helps!

**Comment by benminer** [comment]
@Pagebakers @jdegger Are you all working on getting this working again? I was going to start a fresh project for this with the newest TwilioVoice SDK and CallKit, but if you all are going to work on this one I am happy to help in any way!

**Comment by Pagebakers** [comment]
It seems to be working just fine, but upgrading React with this module seems to give headaches every single time :)

**Comment by jdegger** [comment]
> @Pagebakers @jdegger Are you all working on getting this working again? I was going to start a fresh project for this with the newest TwilioVoice SDK and CallKit, but if you all are going to work on this one I am happy to help in any way! 

By all means, if you are willing to dedicate time on creating a new react native plugin with the newer sdks, it would never hurt! As mentioned, I'm no maintainer on this plugin and not actively working on development. We however, have posted some forks which showcase how our callkeep integration works. 

If the project get serious and stable, we are willing to sponsor the development as well.

**Comment by benminer** [comment]
> It seems to be working just fine, but upgrading React with this module seems to give headaches every single time :)

What upgrades are you trying to perform? To which version?

**Comment by benminer** [comment]
> > @Pagebakers @jdegger Are you all working on getting this working again? I was going to start a fresh project for this with the newest TwilioVoice SDK and CallKit, but if you all are going to work on this one I am happy to help in any way!
> 
> By all means, if you are willing to dedicate time on creating a new react native plugin with the newer sdks, it would never hurt! As mentioned, I'm no maintainer on this plugin and not actively working on development. We however, have posted some forks which showcase how our callkeep integration works.
> 
> If the project get serious and stable, we are willing to sponsor the development as well.

Great to hear. I am also on a smaller scale project heavily relying on Twilio and everything broke with iOS 14. I am going to try out this PR, and perhaps if we get it working we can fork the project to a fresh repository?

**Comment by liemdo** [comment]
Since the maintainer of this project cannot be reached out, I think that it's best to fork to a fresh repo. I don't have much experience with native app development but I can definitely contribute to the new project with the React Native side. We use Twilio heavily on the browser but just starting working on the mobile app lately.

**Comment by benminer** [comment]
> Since the maintainer of this project cannot be reached out, I think that it's best to fork to a fresh repo. I don't have much experience with native app development but I can definitely contribute to the new project with the React Native side. We use Twilio heavily on the browser but just starting working on the mobile app lately.

Awesome, glad to hear there is some support. However, if I understand it correctly, I think most of the work needed to be done is on the native side itself, specifically the wrapper classes that link the TwilioVoice SDK functions to the RCT Bridge for React Native use. I am decently experienced with iOS and can handle that, but may need assistance on the Android side. 

Albeit, this is only if I find this PR to not be fixable due to the issues @Pagebakers has raised. I will be tinkering later tonight, I will keep updates posted here.

**Comment by Pagebakers** [comment]
I'm running with 0.63.3 now. Got it working with !use_frameworks in the pod file and not using flipper in development, but archiving is failing for IOS.

I expect the upgrade to newer Twilio sdk versions wont be a big issue.

**Comment by benminer** [comment]
> I'm running with 0.63.3 now. Got it working with !use_frameworks in the pod file and not using flipper in development, but archiving is failing for IOS.
> 
> I expect the upgrade to newer Twilio sdk versions wont be a big issue.

Is this PR up to date with your most recent changes? Would love to pull down your state of things and try it out with a newer Twilio SDK.

**Comment by mysport12** [comment]
I too am using use_frameworks! with a pre-install function for the static libraries.

Podfile we are using:

require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'
platform :ios, '10.0'

$dynamic_frameworks = ['TwilioVoice']

pre_install do |installer|
  installer.pod_targets.each do |pod|
    if !$dynamic_frameworks.include?(pod.name)
      puts "Overriding the static_framework? method for #{pod.name}"
      def pod.build_type;
        Pod::BuildType.static_library
      end
    end
  end
end

target 'APP' do
  use_frameworks!
  config = use_native_modules!
  permissions_path = '../node_modules/react-native-permissions/ios'

  pod 'TwilioVoice', '~> 5.5.1'
  
  ...OTHER PODS

  use_react_native!(:path => config["reactNativePath"])
  
  target 'APPTests' do
    inherit! :complete
    # Pods for testing
  end
  
	use_flipper!
  post_install do |installer|
    flipper_post_install(installer)
  end
end

target 'APP-tvOS' do
  # Pods for APP-tvOS

  target 'APP-tvOSTests' do
    inherit! :search_paths
    # Pods for testing
  end
end

**Comment by jdegger** [comment]
> > Since the maintainer of this project cannot be reached out, I think that it's best to fork to a fresh repo. I don't have much experience with native app development but I can definitely contribute to the new project with the React Native side. We use Twilio heavily on the browser but just starting working on the mobile app lately.
> 
> 
> 
> Awesome, glad to hear there is some support. However, if I understand it correctly, I think most of the work needed to be done is on the native side itself, specifically the wrapper classes that link the TwilioVoice SDK functions to the RCT Bridge for React Native use. I am decently experienced with iOS and can handle that, but may need assistance on the Android side. 
> 
> 
> 
> Albeit, this is only if I find this PR to not be fixable due to the issues @Pagebakers has raised. I will be tinkering later tonight, I will keep updates posted here.

Might I also suggest that any new work could be depending on callkeep for handeling the call invites? This has been working wonders for us on Android.

We know iOS14 is killing apps who do not respond to invites in the correct way as well. Callkeep has this all figured out. 

I'm on my phone now, but somewhere in this comments section I posted a link to our current fork or this PR which includes callkeep on Android and fixes a problem I found with the proximity sensor as well.

**Comment by jdegger** [comment]
Ps: we have a working iOS14 build/archive running. This should also be on the same fork. I think it's auto linked.

**Comment by Pagebakers** [comment]
@benminer Yeah i'm still using the version in this PR, there are no other changes as far as I know.

Also using the latest 5.5.1 sdk version in your POD file should would with this PR, there shouldn't be any breaking changes. So you can run IOS 14.

I quickly looked into upgrading to Android SDK 5, doesn't seem too hard. I might give it a go.

@mysport12 Thanks, i'll try this tomorrow.

**Comment by benminer** [comment]
> Ps: we have a working iOS14 build/archive running. This should also be on the same fork. I think it's auto linked.

Where is this fork? Is it your repo?

**Comment by jdegger** [comment]
> > Ps: we have a working iOS14 build/archive running. This should also be on the same fork. I think it's auto linked.
> 
> 
> 
> Where is this fork? Is it your repo?

Our project is closed source. But our fork of this PR is public on the @x-guard organisation.

**Comment by benminer** [comment]
> > > Ps: we have a working iOS14 build/archive running. This should also be on the same fork. I think it's auto linked.
> > 
> > 
> > Where is this fork? Is it your repo?
> 
> Our project is closed source. But our fork of this PR is public on the @X-Guard organisation.
> 
> Found the link on my phone:
> 
> [android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46](https://github.com/hoxfon/react-native-twilio-programmable-voice/compare/android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46)

Thanks so much! So from what I can gather, this is working as expected on iOS 14 and Android with ^this PR's fixes? iOS needs some version of CallKeep like you've done on the android end of things?

**Comment by jdegger** [comment]
> > > > Ps: we have a working iOS14 build/archive running. This should also be on the same fork. I think it's auto linked.
> 
> > > 
> 
> > > 
> 
> > > Where is this fork? Is it your repo?
> 
> > 
> 
> > Our project is closed source. But our fork of this PR is public on the @X-Guard organisation.
> 
> > 
> 
> > Found the link on my phone:
> 
> > 
> 
> > [android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46](https://github.com/hoxfon/react-native-twilio-programmable-voice/compare/android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46)
> 
> 
> 
> Thanks so much! So from what I can gather, this is working as expected on iOS 14 and Android with ^this PR's fixes? iOS needs some mirror of CallKeep like you've done on the android end of things? 

Yes, Android and callkeep work find in this version. 

iOS builds fine and outgoing calls are great too. Incoming calls still have some problems like missing interfaces. We hope callkeep can resolve this. 

If the work we did for Android (which was admittedly taken from suggestions made on this repo too, not all our work!) is mirrored to iOS. It should be fully featured once again.

**Comment by benminer** [comment]
> > > > > Ps: we have a working iOS14 build/archive running. This should also be on the same fork. I think it's auto linked.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > Where is this fork? Is it your repo?
> > 
> > 
> > > 
> > 
> > 
> > > Our project is closed source. But our fork of this PR is public on the @X-Guard organisation.
> > 
> > 
> > > 
> > 
> > 
> > > Found the link on my phone:
> > 
> > 
> > > 
> > 
> > 
> > > [android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46](https://github.com/hoxfon/react-native-twilio-programmable-voice/compare/android-twilio-v3...X-Guard:85e4152c91a752e63d9a47682f9f347f00177f46)
> > 
> > 
> > Thanks so much! So from what I can gather, this is working as expected on iOS 14 and Android with ^this PR's fixes? iOS needs some mirror of CallKeep like you've done on the android end of things?
> 
> Yes, Android and callkeep work find in this version.
> 
> iOS builds fine and outgoing calls are great too. Incoming calls still have some problems like missing interfaces. We hope callkeep can resolve this.
> 
> If the work we did for Android (which was admittedly taken from suggestions made on this repo too, not all our work!) is mirrored to iOS. It should be fully featured once again.

Sweet, so if I get that working somehow on your private fork, you think X-Guard could publish it to NPM considering this repo is dead?

**Comment by jdegger** [comment]
Sure! If it works we would be very happy too. 

It's 10:21 PM here so I'm going to bed now. Let me know what you plan to do and you can also reach out to me privately of course. Thanks for your enthusiasm so far!

**Comment by fabriziomoscon** [comment]
I don't have time to answer to this right now. But I will try to have a look later on.
The call for maintainers is always open, I simply never had any volunteer.

**Comment by fabriziomoscon** [comment]
I use this branch in PROD, both iOS and Android
```
// package.json
"react": "16.9.0",
"react-native": "0.61.5",
"react-native-twilio-programmable-voice": "https://github.com/hoxfon/react-native-twilio-programmable-voice#android-twilio-v3",
```

**Comment by jdegger** [comment]
@fabriziomoscon We are willing to take on the maintainer roll now. Last time you asked we declined as we are not sufficient iOS and Android developers. However we think that can help and work moving this repo forward by reviewing and accepting PR's etc.

For example if @benminer sends in a PR for this branch we can test it.

If we take on this role though, are you staying on as the maintainer/owner? How would we decide on a direction of this plugin?

For example: I strongly feel that depending on CallKeep and implementing the CallKeep API in this plugin would be very beneficial. However, this is a rather large change of API. A design decision like that has a big influence on how to proceed. 

Let me know. If you don't have to time discuss it, and trust us as it is without any prior discussion you can of course just make me maintainer for now.

Your efforts on this repo are still greatly appreciated.

**Comment by Pagebakers** [comment]
@jdegger this did the trick, thanks for the help!

```
$dynamic_frameworks = ['TwilioVoice']

pre_install do |installer|
installer.pod_targets.each do |pod|
if !$dynamic_frameworks.include?(pod.name)
puts "Overriding the static_framework? method for #{pod.name}"
def pod.build_type;
Pod::BuildType.static_library
end
end
end
end
```

**Comment by jdegger** [comment]
I stumbled upon this new repo as well, might be interesting.

https://github.com/MrHertal/react-native-twilio-phone

@mrhertal

**Comment by fabriziomoscon** [comment]
Hi @jdegger I propose the following.
- I will merge this PR, tag and release on npm, as I see that most people got it working and it is a stepping stone for implementing newer APIs.
- I will contact you privately to setup a call to discuss maintenance going forward.
- We can open a new issue to discuss how to implement CallKeep

**Comment by jdegger** [comment]
@fabriziomoscon great!

**Comment by jdegger** [comment]
@mrhental are you available to discuss merging your package / knowledge into this one? It would be beneficial to the react native ecosystem to see if we can collaborate on the same package and share knowledge instead of 'compete' (not really competing but you catch my drift).

**Comment by fabriziomoscon** [comment]
`react-native-twilio-programmable-voice@4.0.0`

**Comment by jdegger** [comment]
@MrHertal are you available to discuss merging your package / knowledge into this one? It would be beneficial to the react native ecosystem to see if we can collaborate on the same package and share knowledge instead of 'compete' (not really competing but you catch my drift).

**Comment by MrHertal** [comment]
Hi @jdegger, sorry for late reply. To be honest, I thought a lot about what to do before starting that new project. I am currently working on a React Native app that must include VoIP calls with Twilio and I don't have much time left to finish it.

I understand that you are working on transforming that library to use CallKeep. I agree this is the best option. But I also thought it would take more time upgrading this library than starting a new project.

This library is originally an "all-in-one" package inspired by Twilio quick-start projects. It includes notifications handling, calls workflow and also the UI. In my opinion, the best thing to do is to split this package to let CallKeep manage the UI. It should then be split again to let `@react-native-firebase/messaging` and `react-native-voip-push-notification` manage the notifications, because those libraries are specifically focusing on this task and are well maintained.

Splitting this package like that requires more work than just starting a new project that is relying on mentioned librairies and that only manage the communication with Twilio (start call, hold call, disconnect call...). That is the reason I made https://github.com/MrHertal/react-native-twilio-phone

Now that it is released I can move on my app. If you want to use it in any ways in this repo, you are free to do it. Hope you understand that this is not about competing, but finding the most efficient way to work.

**Comment by benminer** [comment]
@MrHertal This is awesome! Thanks so much for doing this!

**Comment by jdegger** [comment]
@MrHertal No worries and thanks for your elaboration, not that you owed us one! 

We fully understand your motivation for the package. I have a call scheduled tonight with @fabriziomoscon to discuss how we are going to proceed with this repo. If by any chance you have time left @MrHertal to help with this library as you obviously have the knowledge on how to do so :), it would be very welcome.

**Comment by jdegger** [comment]
This discussion has been moved to #158, we invite you to engage and read our update 👍

**Pull request merged**